### PR TITLE
add user data for cookies for marketing

### PR DESCRIPTION
**Pull request by maureensturgeon**
The changes here add cookies which can be used by marketing to segment the user experience on Optimizely. The cookies will be added when the teacher visits the homepage and will be removed on log-out. I was able to verify that the cookies were being set through tests and by running locally.

## Links

- jira ticket: [LP-2039](https://codedotorg.atlassian.net/browse/LP-2039)

## Testing story
Tested through unit tests that set-cookies header had expected values

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -2342,8 +2350,51 @@ def increment_section_attempts
...
+    }.compact
+  end
+
+  def self.marketing_segment_data_keys
+    %w(locale account_age_in_years grades courses has_attended_pd within_usschool_percent_frl school_title_i)
+  end
+
   private
 
+  def account_age_in_years
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb:2378]: Just checking that we always want to round down here?

File: dashboard/app/models/user.rb [hunk 2]
```diff
@@ -2342,8 +2350,51 @@ def increment_section_attempts
...
+    ((Time.now - created_at.to_time) / 1.year).floor
+  end
+
+  # Returns a list of all grades that the teacher currently has sections for
+  def grades_being_taught
+    sections.map(&:grade).uniq
+  end
+
+  # Returns a list of all courses that the teacher currently has sections for
+  def courses_being_taught
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Should we call this something like `curriculums_being_taught` or something similar?  (I'm worried that `course` means something specific in the model.)

File: dashboard/app/models/user.rb [hunk 3]
```diff
@@ -2342,8 +2350,51 @@ def increment_section_attempts
...
+      courses: courses_being_taught.any? ? courses_being_taught.to_json : nil,
+      has_attended_pd: has_attended_pd?,
+      within_us: within_united_states?,
+      school_percent_frl: school_stats&.frl_eligible_total,
+      school_title_i: school_stats&.title_i_status
+    }.compact
+  end
+
+  def self.marketing_segment_data_keys
+    %w(locale account_age_in_years grades courses has_attended_pd within_usschool_percent_frl school_title_i)
```
Comment (jamescodeorg) [inline dashboard/app/models/user.rb]: Should we add a test that checks that this list is in sync with the data returned above?

File: dashboard/test/controllers/api_controller_queries_test.rb [hunk 4]
```diff
@@ -23,7 +23,7 @@ def setup
 
     sign_in_as section.teacher
 
-    assert_queries 11 do
+    assert_queries 15 do
```
Comment (jamescodeorg) [inline dashboard/test/controllers/api_controller_queries_test.rb]: Are we worried about all of these query counts going up by so much? Do we incur this cost on every request or just login?

**Review by jamescodeorg — COMMENTED** [review]
Nice work!

Did you mean to omit schema_cache.yml from this PR?

File: dashboard/app/models/user.rb (hunk 1 above)
Comment (maureensturgeon) [reply dashboard/app/models/user.rb:2378]: I was on the fence about this, I'll change it to `.round`

File: dashboard/test/controllers/api_controller_queries_test.rb (hunk 4 above)
Comment (maureensturgeon) [reply dashboard/test/controllers/api_controller_queries_test.rb]: These requests are made in `after_set_user`, so it should just be when the user logs in. It follows the pattern for setting cookies at login like the experiments cookie. It's not ideal that I've added extra queries, but I didn't see a great alternative.

Comment (maureensturgeon) [reply dashboard/test/controllers/api_controller_queries_test.rb]: I moved where the cookies get set to the homepage which eliminated the extra queries.

**Review by jamescodeorg — APPROVED** [review]
🎉

**Comment by breville** [comment]
Very interesting integration with Optimizely!  Just curious about a couple things:

- Is there any concern about exposing this particular data via non-HttpOnly cookies?  (Understood that Optimizely needs them to be accessible by the client.)
- This adds quite a few cookies, which we need to track for policy reasons (cc @dju90).  Would it be possible to put all of this into a single cookie?  I think we could then give Optimizely some JS to parse this.

**Comment by maureensturgeon** [comment]
> Very interesting integration with Optimizely! Just curious about a couple things:
> 
> * Is there any concern about exposing this particular data via non-HttpOnly cookies?  (Understood that Optimizely needs them to be accessible by the client.)

Before I added these cookies, I confirmed with Travis that there are no data privacy issues with adding these cookies.

> * This adds quite a few cookies, which we need to track for policy reasons (cc @dju90).  Would it be possible to put all of this into a single cookie?  I think we could then give Optimizely some JS to parse this.

The benefits of having the data in separate cookies is it makes it easy for marketing to set up their own custom audiences, but I think that you're right, it'd be possible to combine these into one cookie, it would just take a bit of additional work/guidance for marketing to set up new audiences.

**Pull request merged**